In [2]:
import pandas as pd
import numpy as np

In [3]:
## mount the file / upload the .json file before running all the below codes

df = pd.read_json("retention.json")
df.head()

0                  1                  2      \
gender                      M                  M                  M   
age                  21 years           23 years           24 years   
gross_income       $ 79070.91        $ 178270.68         $ 31243.56   
segment           Individuals  College_Graduated  College_Graduated   
num_credit_cards            0                  0                  0   

                              3            4            5            6      \
gender                            M            M            F            F   
age                        24 years     31 years     32 years     43 years   
gross_income            $ 130740.54  $ 112975.17   $ 37318.56   $ 131281.5   
segment           College_Graduated  Individuals  Individuals  Individuals   
num_credit_cards                  0            0            0            0   

                        7            8                  9      ...  \
gender                      M            F                  M  ...   
age                  48 years     36 years           26 years  ...   
gross_income       $ 184641.6  $ 164347.86         $ 27311.31  ...   
segment           Individuals          VIP  College_Graduated  ...   
num_credit_cards            0            0                  0  ...   

                        54020              54021        54022        54023  \
gender                      F                  M            F            F   
age                  66 years           24 years     65 years     46 years   
gross_income      $ 214340.37         $ 68385.87  $ 266600.04   $ 23037.99   
segment           Individuals  College_Graduated  Individuals  Individuals   
num_credit_cards            0                  0            1            0   

                              54024              54025        54026  \
gender                            M                  M            F   
age                        24 years           24 years     46 years   
gross_income              $ 75405.6         $ 47237.01  $ 106930.05   
segment           College_Graduated  College_Graduated  Individuals   
num_credit_cards                  0                  0            1   

                              54027              54028              54029  
gender                            M                  F                  M  
age                        24 years           27 years           20 years  
gross_income            $ 185323.05         $ 65246.79        $ 180678.48  
segment           College_Graduated  College_Graduated  College_Graduated  
num_credit_cards                  0                  0                  0  

[5 rows x 54030 columns]

In [4]:
## converting the json to csv

df.to_csv("retention.csv")

In [5]:
## transposing data and keeping the first column as index 

pd.read_csv('retention.csv',header = None).T.to_csv('retention_updated.csv',header = False, index = False)

retention = pd.read_csv('retention_updated.csv',index_col=0)

In [6]:
## following codes are basic inspection and to ensure the data is consistent
retention.columns

Index(['gender', 'age', 'gross_income', 'segment', 'num_credit_cards',
       'tiprel_1mes', 'ind_actividad_cliente', 'num_Products', 'num_loans',
       'duration'],
      dtype='object')

In [7]:
## inspection
retention.index

Int64Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,
                9,
            ...
            54020, 54021, 54022, 54023, 54024, 54025, 54026, 54027, 54028,
            54029],
           dtype='int64', length=54030)

In [8]:
## inspection
retention.values

array([['M', '21 years', '$ 79070.91', ..., 1, 0, 1035],
       ['M', '23 years', '$ 178270.68', ..., 1, 0, 1097],
       ['M', '24 years', '$ 31243.56', ..., 1, 0, 1866],
       ...,
       ['M', '24 years', '$ 185323.05', ..., 0, 0, 1817],
       ['F', '27 years', '$ 65246.79', ..., 1, 0, 1447],
       ['M', '20 years', '$ 180678.48', ..., 1, 0, 366]], dtype=object)

In [9]:
## inspection - counting the total columns and rows
retention.shape

(54030, 10)

In [10]:
## checking how the data looks by pulling the top 10 rows
retention.head(10) 

gender       age gross_income            segment  num_credit_cards  \
0      M  21 years   $ 79070.91        Individuals                 0   
1      M  23 years  $ 178270.68  College_Graduated                 0   
2      M  24 years   $ 31243.56  College_Graduated                 0   
3      M  24 years  $ 130740.54  College_Graduated                 0   
4      M  31 years  $ 112975.17        Individuals                 0   
5      F  32 years   $ 37318.56        Individuals                 0   
6      F  43 years   $ 131281.5        Individuals                 0   
7      M  48 years   $ 184641.6        Individuals                 0   
8      F  36 years  $ 164347.86                VIP                 0   
9      M  26 years   $ 27311.31  College_Graduated                 0   

  tiprel_1mes  ind_actividad_cliente  num_Products  num_loans  duration  
0           I                      0             1          0      1035  
1           I                      0             1          0      1097  
2           I                      1             1          0      1866  
3           I                      0             1          0      1440  
4           I                      0             0          0       376  
5           A                      1             4          0       699  
6           A                      0             1          0      5022  
7           A                      1             1          0      4066  
8           A                      1             1          0      1427  
9           I                      1             1          0      1458

In [11]:
## describe by default I think takes only numeric value. I have included all just to check how the data looks
retention.describe(include = 'all')

gender       age gross_income      segment  num_credit_cards  \
count   54030     54030        54030        54030      54030.000000   
unique      2       110        49994            3               NaN   
top         F  23 years  $ 451931.22  Individuals               NaN   
freq    29681      3080           29        32607               NaN   
mean      NaN       NaN          NaN          NaN          0.049880   
std       NaN       NaN          NaN          NaN          0.217698   
min       NaN       NaN          NaN          NaN          0.000000   
25%       NaN       NaN          NaN          NaN          0.000000   
50%       NaN       NaN          NaN          NaN          0.000000   
75%       NaN       NaN          NaN          NaN          0.000000   
max       NaN       NaN          NaN          NaN          1.000000   

       tiprel_1mes  ind_actividad_cliente  num_Products     num_loans  \
count        54030           54030.000000  54030.000000  54030.000000   
unique           2                    NaN           NaN           NaN   
top              I                    NaN           NaN           NaN   
freq         29026                    NaN           NaN           NaN   
mean           NaN               0.469073      1.536017      0.009365   
std            NaN               0.499047      1.578848      0.096320   
min            NaN               0.000000      0.000000      0.000000   
25%            NaN               0.000000      1.000000      0.000000   
50%            NaN               0.000000      1.000000      0.000000   
75%            NaN               1.000000      2.000000      0.000000   
max            NaN               1.000000     13.000000      1.000000   

            duration  
count   54030.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean     2913.936665  
std      2027.956268  
min       366.000000  
25%      1118.000000  
50%      2039.500000  
75%      4693.000000  
max      7909.000000

In [12]:
## we see here that few of the data that we would want as integer or float is stored as object (age, tiprel_1mes)

retention.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 54030 entries, 0 to 54029
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   gender                 54030 non-null  object
 1   age                    54030 non-null  object
 2   gross_income           54030 non-null  object
 3   segment                54030 non-null  object
 4   num_credit_cards       54030 non-null  int64 
 5   tiprel_1mes            54030 non-null  object
 6   ind_actividad_cliente  54030 non-null  int64 
 7   num_Products           54030 non-null  int64 
 8   num_loans              54030 non-null  int64 
 9   duration               54030 non-null  int64 
dtypes: int64(5), object(5)
memory usage: 4.5+ MB


In [13]:
## checking for any na
retention.isna().sum()

gender                   0
age                      0
gross_income             0
segment                  0
num_credit_cards         0
tiprel_1mes              0
ind_actividad_cliente    0
num_Products             0
num_loans                0
duration                 0
dtype: int64

In [14]:
## checking for null values
retention.isnull().sum()

gender                   0
age                      0
gross_income             0
segment                  0
num_credit_cards         0
tiprel_1mes              0
ind_actividad_cliente    0
num_Products             0
num_loans                0
duration                 0
dtype: int64

In [15]:
## replace "years" with "" in the age
## similarly replacing "$" with ""


retention['age'] = retention['age'].str.replace('years','',regex = True)
retention['gross_income'] = retention['gross_income'].str.replace('$','',regex = True)

## renaming columns that had unclear column names
## tiprel_1mes is renamed to inactivity_begn
## ind_actividad_cliente is renamed to inactivity_end

retention.rename(columns={'tiprel_1mes':'inactivity_begn'},inplace = True)
retention.rename(columns={'ind_actividad_cliente':'inactivity_end'}, inplace = True)

retention.head()

gender  age gross_income            segment  num_credit_cards  \
0      M  21      79070.91        Individuals                 0   
1      M  23     178270.68  College_Graduated                 0   
2      M  24      31243.56  College_Graduated                 0   
3      M  24     130740.54  College_Graduated                 0   
4      M  31     112975.17        Individuals                 0   

  inactivity_begn  inactivity_end  num_Products  num_loans  duration  
0               I               0             1          0      1035  
1               I               0             1          0      1097  
2               I               1             1          0      1866  
3               I               0             1          0      1440  
4               I               0             0          0       376

In [16]:
## to make the change permanent, pulled it to another csv
## need to check if there is a better / faster way to do this

retention.to_csv('df1.csv')


In [17]:
df1 = pd.read_csv('df1.csv',index_col=0)
df1.head(2)

## converting the flag to 0 and 1 for the inacitivty at the beginning
df1.inactivity_begn[df1.inactivity_begn =='I'] = 0
df1.inactivity_begn[df1.inactivity_begn =='A'] = 1 

df1.head()


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


gender  age  gross_income            segment  num_credit_cards  \
0      M   21      79070.91        Individuals                 0   
1      M   23     178270.68  College_Graduated                 0   
2      M   24      31243.56  College_Graduated                 0   
3      M   24     130740.54  College_Graduated                 0   
4      M   31     112975.17        Individuals                 0   

  inactivity_begn  inactivity_end  num_Products  num_loans  duration  
0               0               0             1          0      1035  
1               0               0             1          0      1097  
2               0               1             1          0      1866  
3               0               0             1          0      1440  
4               0               0             0          0       376

In [18]:
## converting to df2
df1.to_csv('df2.csv')

In [19]:
## readiing the df2 csv. This will be the final file we will be using
df2 = pd.read_csv('df2.csv')

## checking if the data is as per our desired format

df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54030 entries, 0 to 54029
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        54030 non-null  int64  
 1   gender            54030 non-null  object 
 2   age               54030 non-null  int64  
 3   gross_income      54030 non-null  float64
 4   segment           54030 non-null  object 
 5   num_credit_cards  54030 non-null  int64  
 6   inactivity_begn   54030 non-null  int64  
 7   inactivity_end    54030 non-null  int64  
 8   num_Products      54030 non-null  int64  
 9   num_loans         54030 non-null  int64  
 10  duration          54030 non-null  int64  
dtypes: float64(1), int64(8), object(2)
memory usage: 4.5+ MB


The below dataset clearly shows that the data has discrepancies. Min age of a customer is 2 and the max age is 112.

Further validation of the data is necessary  to ensure we are looking at it correctly

---



In [47]:
## key stats for our final data set
df1.describe().round(2)

age  gross_income  num_credit_cards  inactivity_end  num_Products  \
count  54030.00      54030.00          54030.00        54030.00      54030.00   
mean      40.77     134196.67              0.05            0.47          1.54   
std       17.21     239721.46              0.22            0.50          1.58   
min        2.00       2336.01              0.00            0.00          0.00   
25%       25.00      68581.07              0.00            0.00          1.00   
50%       40.00     101566.32              0.00            0.00          1.00   
75%       51.00     156253.76              0.00            1.00          2.00   
max      112.00   28894395.51              1.00            1.00         13.00   

       num_loans  duration  
count   54030.00  54030.00  
mean        0.01   2913.94  
std         0.10   2027.96  
min         0.00    366.00  
25%         0.00   1118.00  
50%         0.00   2039.50  
75%         0.00   4693.00  
max         1.00   7909.00

In [21]:
## calculating the mean and rounding it to 2 decimals
df2[['gross_income','num_credit_cards','num_Products','num_loans','duration']].mean().round(2)

gross_income        134196.67
num_credit_cards         0.05
num_Products             1.54
num_loans                0.01
duration              2913.94
dtype: float64

In [22]:
## calculating the median and rounding it to 2 decimals
df2[['gross_income','num_credit_cards','num_Products','num_loans','duration']].median().round(2)

gross_income        101566.32
num_credit_cards         0.00
num_Products             1.00
num_loans                0.00
duration              2039.50
dtype: float64

In [23]:
## calculating the mode and rounding it to 2 decimals
df2[['gross_income','num_credit_cards','num_Products','num_loans','duration']].mode().round(2)

gross_income  num_credit_cards  num_Products  num_loans  duration
0     451931.22                 0             1          0       716

In [24]:
## calculating the quantiles and rounding it to 2 decimals
df2[['gross_income','num_credit_cards','num_Products','num_loans','duration']].quantile([0.25, 0.5, 0.75, 0.90, 0.95]).round(2)

gross_income  num_credit_cards  num_Products  num_loans  duration
0.25      68581.07               0.0           1.0        0.0    1118.0
0.50     101566.32               0.0           1.0        0.0    2039.5
0.75     156253.76               0.0           2.0        0.0    4693.0
0.90     234387.88               0.0           3.0        0.0    5844.0
0.95     310033.37               0.0           5.0        0.0    6529.0

## Data Exploration

Now that the data is cleaned (to the best of my ability), moving to Task 2 of the project

In [29]:
df2.groupby(by = ['gender','segment']).agg({'gross_income' : 'mean'}).round(2)

gross_income
gender segment                        
F      College_Graduated     120272.44
       Individuals           139036.32
       VIP                   171891.73
M      College_Graduated     112403.68
       Individuals           145062.79
       VIP                   194549.85

In [87]:
## creating the pivot table to analyze the data

P1 = pd.pivot_table(df2, index = ['inactivity_begn','inactivity_end','gender'],
                    values = ['age','gross_income','num_credit_cards','num_Products'],
               aggfunc = ['mean','median','count']).round(2)
P1.style.format('{:,}')

Here I am identifying the buckets as follows:

1.   Inactive at begin | inactive at end
2.   Inactive at  begin | active at end
1.   Active at begin | inactive at end
2.   Active at begin | active at end

We see that the customers with the lowest annual income of the 4 buckets have changed their behaviour i.e. inactive to active and they seem to be using 1 product of the bank

In [89]:
P1.query('inactivity_end == 1').style.format('{:,}')

People wih high income tend to be active. They seem to be using more products as well compared to the group that got inactive

Compared to the pivot table above as well it seems that the group with the highest annual income are more active in terms of using banking products. Females seem to be using more products on average compared to Males in this bucket

##Analysis of behaviour

Almost 50% of the customers have not changed their preference (inactive during beginning and at the end)

Post that 42% of the customers have remained active with the bank.

This suggests that over the six months, only 8% of the customers have changed their preference

PS - all % are based on the total count of 54030 customers that are studied in this analysis

In [91]:
P2 = pd.pivot_table(df2, index = ['inactivity_begn','inactivity_end','segment'],
                    values = ['age','gross_income','num_Products'],
               aggfunc = ['mean','median','count']).round(2)
P2.style.format('{:,}')

## Looking at the data from Bank Defined Segments

We see that 31 VIP customers became inactive. However they were not using most of the bank's products and hence do not seem to be a major  risk to the bank based on the data available.

There are significant numbers of individuals with an average income of $141,000 that the bank can target. This is the category that has remained inactive throughout the 6 month period.

